### PR TITLE
Kaa-96: Logs delivery confirmation

### PR DESCRIPTION
**Pull request by ashvayka**

### CODE REVIEW

File: server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/mvp/view/appender/LogAppenderViewImpl.java [hunk 1]
```diff
@@ -182,6 +204,11 @@ protected String getViewTitle() {
     protected String getSubTitle() {
         return Utils.constants.logAppenderDetails();
     }
+    
+	@Override
+	public HasValue<Boolean> getConfirmDelivery() {
```
Comment (ghost) [inline server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/mvp/view/appender/LogAppenderViewImpl.java]: Different indentation.

File: server/appenders/cdap/appender/src/main/java/org/kaaproject/kaa/server/appenders/cdap/appender/CdapLogAppender.java [hunk 2]
```diff
@@ -65,24 +67,31 @@ public CdapLogAppender() {
...
                 final String appToken = this.getApplicationToken();
                 LOG.debug("[{}] appending {} logs to cdap stream", this.getApplicationToken(), logEventPack.getEvents()
                         .size());
-                for (LogEventDto dto : generateLogEvent(logEventPack, header)) {
-                    StringBuilder sb = new StringBuilder();
-                    sb.append("{\"header\":").append(dto.getHeader()).append(",");
-                    sb.append("\"event\":").append(dto.getEvent()).append("}");
-                    ListenableFuture<Void> result = streamWriter.write(sb.toString(), UTF8);
-                    Futures.addCallback(result, new Callback(appToken), callbackExecutor);
+                try{
```
Comment (ghost) [inline server/appenders/cdap/appender/src/main/java/org/kaaproject/kaa/server/appenders/cdap/appender/CdapLogAppender.java]: Need code style spaces both in 'try' and 'catch' statements.

File: server/appenders/cdap/appender/src/test/java/org/kaaproject/kaa/server/appenders/cdap/appender/CdapLogAppenderTest.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
+package org.kaaproject.kaa.server.appenders.cdap.appender;
```
Comment (ghost) [inline server/appenders/cdap/appender/src/test/java/org/kaaproject/kaa/server/appenders/cdap/appender/CdapLogAppenderTest.java:16]: Need to add license header.

File: server/appenders/cdap/appender/src/test/java/org/kaaproject/kaa/server/appenders/cdap/appender/CdapLogAppenderTest.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    
+    @Test
+    public void testAppend() throws IOException{
+        LogEventPack logEventPack = getLogEventPack();
+        
+        Mockito.when(writer.write(Mockito.anyString(), Mockito.any(Charset.class))).thenReturn(new DummyFuture());
+        
+        LogDeliveryCallback callback = Mockito.mock(LogDeliveryCallback.class);
+        appender.doAppend(logEventPack, callback);
+        Mockito.verify(callback, Mockito.timeout(10000)).onSuccess();
```
Comment (ghost) [inline server/appenders/cdap/appender/src/test/java/org/kaaproject/kaa/server/appenders/cdap/appender/CdapLogAppenderTest.java]: Maybe better to extract hardcoded timeout to some constant?

File: server/appenders/cdap/appender/src/test/java/org/kaaproject/kaa/server/appenders/cdap/appender/CdapLogAppenderTest.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    @Test
+    public void testAppendToClosed() throws IOException{
+        LogEventPack logEventPack = getLogEventPack();
+        appender.close();
+        LogDeliveryCallback callback = Mockito.mock(LogDeliveryCallback.class);
+        appender.doAppend(logEventPack, callback);
+        Mockito.verify(callback, Mockito.timeout(10000)).onInternalError();
+    }
+
+	private LogEventPack getLogEventPack() throws IOException {
```
Comment (ghost) [inline server/appenders/cdap/appender/src/test/java/org/kaaproject/kaa/server/appenders/cdap/appender/CdapLogAppenderTest.java]: Different indentation.

File: server/appenders/file/appender/src/main/java/org/kaaproject/kaa/server/appenders/file/appender/FileSystemLogAppender.java [hunk 6]
```diff
@@ -49,18 +51,25 @@ public FileSystemLogAppender() {
...
-    public void doAppend(LogEventPack logEventPack, RecordHeader header) {
+    public void doAppend(LogEventPack logEventPack, RecordHeader header, LogDeliveryCallback listener) {
         if (!closed) {
-            String path = logsRootPath + "/" + tenantDirName + "/" + applicationDirName;
-            LOG.debug("[{}] appending {} logs to directory", path, logEventPack.getEvents().size());
-            List<String> dtos = eventsToStrings(generateLogEvent(logEventPack, header));
-            LOG.debug("[{}] saving {} objects", path, dtos.size());
-            for(String event : dtos){
-                logger.append(event);
+            try{
```
Comment (ghost) [inline server/appenders/file/appender/src/main/java/org/kaaproject/kaa/server/appenders/file/appender/FileSystemLogAppender.java]: Need to add space symbol.

File: server/common/dao/src/main/java/org/kaaproject/kaa/server/common/dao/model/sql/LogAppender.java [hunk 7]
```diff
@@ -158,7 +163,15 @@ public void setMaxLogSchemaVersion(int maxLogSchemaVersion) {
         this.maxLogSchemaVersion = maxLogSchemaVersion;
     }
 
-    public LogAppenderStatusDto getStatus() {
+    public boolean isConfirmDelivery() {
+		return confirmDelivery;
+	}
+
+	public void setConfirmDelivery(boolean confirmDelivery) {
```
Comment (ghost) [inline server/common/dao/src/main/java/org/kaaproject/kaa/server/common/dao/model/sql/LogAppender.java]: Different indentation.

File: server/common/dto/src/main/java/org/kaaproject/kaa/common/dto/logs/LogAppenderBaseDto.java [hunk 8]
```diff
@@ -0,0 +1,265 @@
+package org.kaaproject.kaa.common.dto.logs;
```
Comment (ghost) [inline server/common/dto/src/main/java/org/kaaproject/kaa/common/dto/logs/LogAppenderBaseDto.java:16]: Need to add license header.

File: server/appenders/cdap/appender/src/test/java/org/kaaproject/kaa/server/appenders/cdap/appender/CdapLogAppenderTest.java (hunk 3 above)
Comment (ashvayka) [inline server/appenders/cdap/appender/src/test/java/org/kaaproject/kaa/server/appenders/cdap/appender/CdapLogAppenderTest.java:16]: done

File: server/common/dto/src/main/java/org/kaaproject/kaa/common/dto/logs/LogAppenderBaseDto.java (hunk 8 above)
Comment (ashvayka) [inline server/common/dto/src/main/java/org/kaaproject/kaa/common/dto/logs/LogAppenderBaseDto.java:16]: done

File: server/operations/src/test/java/org/kaaproject/kaa/server/operations/service/akka/messages/core/logs/LogEventPackMessageTest.java [hunk 9]
```diff
@@ -26,50 +26,52 @@
...
+
     @Test
     public void logEventPackDataTest() {
         LogEventPack logEventPack = new LogEventPack(ENDPOINT_KEY, DATE_CREATED, LOG_SCHEMA, EVENTS);
         logEventPack.setLogSchemaVersion(LOG_SCHEMA_VERSION);
-        
-        LogEventPackMessage logEvent = new LogEventPackMessage(logEventPack);
-        
+
+        LogEventPackMessage logEvent = new LogEventPackMessage("request1", ActorRef.noSender(), logEventPack);
```
Comment (ikulikov) [inline server/operations/src/test/java/org/kaaproject/kaa/server/operations/service/akka/messages/core/logs/LogEventPackMessageTest.java]: "request1" can be defined as string constant

File: server/operations/src/test/java/org/kaaproject/kaa/server/operations/service/logs/DefaultLogAppenderBuilderTest.java [hunk 10]
```diff
@@ -99,10 +100,16 @@ public void close() {
...
-		}
+        @Override
+        public boolean isSchemaVersionSupported(int version) {
+            // TODO Auto-generated method stub
+            return true;
+        }
+
+        @Override
+        public boolean isDeliveryConfirmationRequired() {
+            // TODO Auto-generated method stub
```
Comment (ikulikov) [inline server/operations/src/test/java/org/kaaproject/kaa/server/operations/service/logs/DefaultLogAppenderBuilderTest.java]: "TODO" comments should be removed

**Pull request merged**